pkg/controller/servicecidrs/servicecidrs_controller.go

```diff
@@ -0,0 +1,641 @@
+	}
+	c.queue.Add(cidr.Name)
+	for _, key := range c.cidrToCidrs(cidr) {
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) updateServiceCIDR(oldObj, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// deleteServiceCIDR
+func (c *Controller) deleteServiceCIDR(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err == nil {
+		c.queue.Add(key)
+	}
+}
+
+// addIPAddress may block a ServiceCIDR deletion
+func (c *Controller) addIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		return
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// deleteIPAddress may unblock a ServiceCIDR deletion
+func (c *Controller) deleteIPAddress(obj interface{}) {
+	ip, ok := obj.(*networkingapiv1alpha1.IPAddress)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			return
+		}
+		ip, ok = tombstone.Obj.(*networkingapiv1alpha1.IPAddress)
+		if !ok {
+			return
+		}
+	}
+
+	for _, cidr := range c.ipToCidrs(ip) {
+		c.queue.Add(cidr)
+	}
+}
+
+// cidrToCidrs, given a ServiceCIDR return the ServiceCIDRs that contain or are contained
+// this is required because adding or removing a CIDR will require to recompute the
+// state of each ServiceCIDR to check if can be unblocked on deletion.
+func (c *Controller) cidrToCidrs(cidr *networkingapiv1alpha1.ServiceCIDR) []string {
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+
+	cidrs := sets.New[string]()
+	if prefix, err := netip.ParsePrefix(cidr.Spec.IPv4); err == nil { // if is empty err will not be nil
+		c.tree.WalkPath(prefix, func(k netip.Prefix, v cidrSet) bool {
+			cidrs.Insert(v.UnsortedList()...)
+			return false
+		})
+		c.tree.WalkPrefix(prefix, func(k netip.Prefix, v cidrSet) bool {
+			cidrs.Insert(v.UnsortedList()...)
+			return false
+		})
+	}
+	if prefix, err := netip.ParsePrefix(cidr.Spec.IPv6); err == nil { // if is empty err will not be nil
+		c.tree.WalkPath(prefix, func(k netip.Prefix, v cidrSet) bool {
+			cidrs.Insert(v.UnsortedList()...)
+			return false
+		})
+		c.tree.WalkPrefix(prefix, func(k netip.Prefix, v cidrSet) bool {
+			cidrs.Insert(v.UnsortedList()...)
+			return false
+		})
+	}
+	return cidrs.UnsortedList()
+}
+
+// ipToCidrs, given an IPAddress return the ServiceCIDRs that contains the IP,
+// as it may block or be blocking the deletion of the ServiceCIDRs that contain it.
+func (c *Controller) ipToCidrs(ip *networkingapiv1alpha1.IPAddress) []string {
+	// only process IPs managed by the kube-apiserver
+	managedBy, ok := ip.Labels[networkingapiv1alpha1.LabelManagedBy]
+	if !ok || managedBy != ipallocator.ControllerName {
+		return []string{}
+	}
+
+	prefix, err := ipAddressToPrefix(ip)
+	if err != nil {
+		// This should not happen, the IPAddress object validates
+		// the name is a valid IPAddress
+		return []string{}
+	}
+
+	c.muTree.Lock()
+	defer c.muTree.Unlock()
+	cidrs := []string{}
```

Comment:
Naming comment applies to variables too. I have a strong expectation that "cidr" refers to CIDR strings (or `*net.IPNet`s or `netip.Prefix`es), not ServiceCIDR object names.